src/core/binary.c

```diff
@@ -0,0 +1,201 @@
+* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+* IN THE SOFTWARE.
+*/
+
+#ifndef JANET_AMALG
+#include "features.h"
+#include <janet.h>
+#include "util.h"
+#endif
+
+static int binary_should_reverse(const Janet *argv, int32_t argc) {
+    JanetKeyword order_kw = janet_getkeyword(argv, argc);
+    if (!janet_cstrcmp(order_kw, "le")) {
+#if JANET_BIG_ENDIAN
+        return 1;
+#endif
+    } else if (!janet_cstrcmp(order_kw, "be")) {
+#if JANET_LITTLE_ENDIAN
+        return 1;
+#endif
+    } else if (!janet_cstrcmp(order_kw, "native")) {
+        return 0;
+    } else {
+        janet_panicf("expected endianness :le, :be or :native, got %v", argv[1]);
+    }
+    return 0;
+}
+
+static void binary_reverse_u32(uint8_t bytes[4]) {
+    uint8_t temp;
+    temp = bytes[3];
+    bytes[3] = bytes[0];
+    bytes[0] = temp;
+    temp = bytes[2];
+    bytes[2] = bytes[1];
+    bytes[1] = temp;
+}
+
+static void binary_reverse_u64(uint8_t bytes[8]) {
+    uint8_t temp;
+    temp = bytes[7];
+    bytes[7] = bytes[0];
+    bytes[0] = temp;
+    temp = bytes[6];
+    bytes[6] = bytes[1];
+    bytes[1] = temp;
+    temp = bytes[5];
+    bytes[5] = bytes[2];
+    bytes[2] = temp;
+    temp = bytes[4];
+    bytes[4] = bytes[3];
+    bytes[3] = temp;
+}
+
+JANET_CORE_FN(cfun_binary_write_uint16,
+              "(binary/write-uint16 buffer order data)",
+              "Write the binary representation of data into the buffer.") {
+    janet_fixarity(argc, 3);
+    JanetBuffer *buffer = janet_getbuffer(argv, 0);
+
+    int reverse = binary_should_reverse(argv, 1);
+
+    union {
+        uint16_t data;
+        uint8_t bytes[2];
+    } u;
+
+    u.data = (uint16_t) janet_getinteger(argv, 2);
+
+    if (reverse) {
+        uint8_t temp = u.bytes[1];
+        u.bytes[1] = u.bytes[0];
+        u.bytes[0] = temp;
+    }
+
+    janet_buffer_push_u16(buffer, *(uint16_t*) u.bytes);
+
+    return argv[0];
+}
+
+JANET_CORE_FN(cfun_binary_write_uint32,
+              "(binary/write-uint32 buffer order data)",
+              "Write the binary representation of data into the buffer.") {
+    janet_fixarity(argc, 3);
+    JanetBuffer *buffer = janet_getbuffer(argv, 0);
+
+    int reverse = binary_should_reverse(argv, 1);
+
+    union {
+        uint32_t data;
+        uint8_t bytes[4];
+    } u;
+
+    u.data = (uint32_t) janet_getinteger(argv, 2);
+
+    if (reverse)
+        binary_reverse_u32(u.bytes);
+
+    janet_buffer_push_u32(buffer, *(uint32_t*) u.bytes);
+
+    return argv[0];
+}
+
+JANET_CORE_FN(cfun_binary_write_uint64,
+              "(binary/write-uint64 buffer order data)",
+              "Write the binary representation of data into the buffer.") {
+    janet_fixarity(argc, 3);
+    JanetBuffer *buffer = janet_getbuffer(argv, 0);
+
+    int reverse = binary_should_reverse(argv, 1);
+
+    union {
+        uint64_t data;
+        uint8_t bytes[8];
+    } u;
+
+    u.data = (uint64_t) janet_getuinteger64(argv, 2);
```

Comment:
Maybe there is a historical reason for this but I found that I couldn't write an 8 byte `uint64_t`. This function checks for the unsigned integer range using `JANET_INTMAX_DOUBLE` which is `9007199254740992.0` or hex `20000000000000` (7 bytes).